File: services/hub/cmd/brain/spawn_transport_test.go
package main

import (
	"context"
	"testing"
)

// A codex spawn that names no transport must land HEADLESS — the same shape the
// desktop picks (apps/desktop/src/main/lib/spawnTransport.ts). This is the leg
// most of the fleet actually takes: `agents.spawn` over the bus, from the MCP
// facade or a dispatched worker, names no transport at all. Before the shared
// resolver, an absent key meant "hybrid" here and "hybrid" in claudemon, so a
// bus-dispatched codex worker came up as a TUI + viewer pair while a locally
// spawned one came up GUI-only.
func TestCodexSpawnDefaultsToStreamOnTheWire(t *testing.T) {
	rec := newRecorder()
	srv := rec.server()
	defer srv.Close()
	reg := newSpawnTestRegistry(t, srv.URL)

	if _, err := reg.handle(context.Background(), "agents.spawn",
		[]byte(`{"provider":"codex","cwd":"/tmp"}`)); err != nil {
		t.Fatal(err)
	}
	managed := rec.calls("/sessions/spawn-managed")
	if len(managed) != 1 {
		t.Fatalf("expected one spawn-managed call, got %d", len(managed))
	}
	if got := managed[0].body["transport"]; got != "stream" {
		t.Errorf("codex spawn with no transport must state 'stream', got %v", got)
	}
}

// ...and an explicit 'pty' still reaches the wire, or the hybrid would be
// unreachable the moment the default flipped. The whole point of forwarding
// BOTH values is that "the user chose hybrid" and "the user said nothing" stop
// being the same request.
func TestCodexExplicitPtyStatesItselfOnTheWire(t *testing.T) {
	rec := newRecorder()
	srv := rec.server()
	defer srv.Close()
	reg := newSpawnTestRegistry(t, srv.URL)

	if _, err := reg.handle(context.Background(), "agents.spawn",
		[]byte(`{"provider":"codex","cwd":"/tmp","transport":"pty"}`)); err != nil {
		t.Fatal(err)
	}
	managed := rec.calls("/sessions/spawn-managed")
	if len(managed) != 1 {
		t.Fatalf("expected one spawn-managed call, got %d", len(managed))
	}
	if got := managed[0].body["transport"]; got != "pty" {
		t.Errorf("an explicit hybrid request must ride the wire as 'pty', got %v", got)
	}
}

// The resolver itself: caller > config > shipped fallback, per harness. The
// per-harness fallbacks differ (claude's headless transport is opt-in per
// install, codex's is the default), which is why one shared "?? 'pty'" was
// wrong for codex.
func TestTransportDefaultResolutionOrder(t *testing.T) {
	for _, tc := range []struct {
		name, provider, requested, want string
		cfg                             map[string]any
	}{
		{name: "explicit wins over config", provider: "codex", requested: "pty", want: "pty",
			cfg: map[string]any{"codex": map[string]any{"transport": "stream"}}},
		{name: "config wins over fallback", provider: "codex", want: "pty",
			cfg: map[string]any{"codex": map[string]any{"transport": "pty"}}},
		// A registry with no config file still SEEDS config_defaults.json, so
		// these two read the shipped values rather than the in-code fallback —
		// which is the point: the JSON and this map must agree.
		{name: "codex ships headless", provider: "codex", want: "stream", cfg: map[string]any{}},
		{name: "claude ships headless too", provider: "claude", want: "stream", cfg: map[string]any{}},
		// An unparseable value is the only way to reach the in-code fallback
		// from here, and it must not become a third session shape.
		{name: "junk codex value falls back to stream", provider: "codex", want: "stream",
			cfg: map[string]any{"codex": map[string]any{"transport": "sideways"}}},
		{name: "junk claude value falls back to pty", provider: "claude", want: "pty",
			cfg: map[string]any{"claude": map[string]any{"transport": "sideways"}}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			// A fresh registry (and config home) per case: save() MERGES, so a
			// value set for one case would survive into the next and quietly
			// turn a fallback assertion into a config one.
			rec := newRecorder()
			srv := rec.server()
			defer srv.Close()
			reg := newSpawnTestRegistry(t, srv.URL)
			if len(tc.cfg) > 0 {
				reg.cfg.save(tc.cfg)
			}
			if got := reg.transportDefault(tc.provider, tc.requested); got != tc.want {
				t.Errorf("transportDefault(%q, %q) = %q, want %q", tc.provider, tc.requested, got, tc.want)
			}
		})
	}
}

File: apps/desktop/src/main/lib/spawnTransport.test.ts
/**
 * The transport default, in one place. Before this module the rule lived as
 * `opts.transport ?? cfg.claude.transport ?? 'pty'` copy-pasted at four spawn
 * entry points, and codex had no configured default at all — an absent
 * transport just meant "hybrid", spelled as the ABSENCE of a key. That is what
 * let the same fleet come up GUI-only on one machine and as TUI+viewer pairs on
 * another.
 *
 * TWIN: services/hub/cmd/brain/handlers.go `transportDefault` — a spawn that
 * never touches the desktop must land on the same shape (pinned there by
 * spawn_transport_test.go).
 */
import { describe, it, expect, vi, beforeEach } from 'vitest';

let mockConfig: Record<string, unknown>;
vi.mock('../services/configService', () => ({
  configService: { getConfig: () => mockConfig },
}));

const { resolveTransport, parseTransport, hasTransportChoice, TRANSPORT_FALLBACK } =
  await import('./spawnTransport');

beforeEach(() => {
  mockConfig = {};
});

describe('resolveTransport', () => {
  it('an explicit request always wins, both directions', () => {
    mockConfig = { codex: { transport: 'stream' }, claude: { transport: 'stream' } };
    expect(resolveTransport('codex', 'pty')).toBe('pty');
    expect(resolveTransport('claude', 'pty')).toBe('pty');
    mockConfig = { codex: { transport: 'pty' }, claude: { transport: 'pty' } };
    expect(resolveTransport('codex', 'stream')).toBe('stream');
    expect(resolveTransport('claude', 'stream')).toBe('stream');
  });

  it('falls back to the harness config when nothing was requested', () => {
    mockConfig = { codex: { transport: 'pty' }, claude: { transport: 'stream' } };
    expect(resolveTransport('codex')).toBe('pty');
    expect(resolveTransport('claude')).toBe('stream');
  });

  // The reason this is per-harness at all: codex ships headless, claude's
  // headless transport is opt-in per install. One shared `?? 'pty'` was the bug.
  it('falls back per harness when config says nothing', () => {
    expect(resolveTransport('codex')).toBe('stream');
    expect(resolveTransport('claude')).toBe('pty');
    expect(TRANSPORT_FALLBACK).toEqual({ claude: 'pty', codex: 'stream' });
  });

  it('treats a junk value — in the request or in config — as unstated', () => {
    mockConfig = { codex: { transport: 'sideways' } };
    expect(resolveTransport('codex')).toBe('stream');
    expect(resolveTransport('codex', 'sideways')).toBe('stream');
    expect(resolveTransport('codex', '')).toBe('stream');
  });

  // opencode/pi have exactly one session shape. Inventing a transport for them
  // would put a key on a payload their adapter does not read.
  it.each(['opencode', 'pi'])('%s has no transport choice at all', (provider) => {
    expect(hasTransportChoice(provider)).toBe(false);
    expect(resolveTransport(provider)).toBeUndefined();
    expect(resolveTransport(provider, 'stream')).toBeUndefined();
  });

  it('parseTransport only admits the two real shapes', () => {
    expect(parseTransport('pty')).toBe('pty');
    expect(parseTransport('stream')).toBe('stream');
    for (const junk of ['', 'PTY', 'hybrid', undefined, null, 0, {}]) {
      expect(parseTransport(junk)).toBeUndefined();
    }
  });
});
